apis/Google.Cloud.Datastore.V1/Google.Cloud.Datastore.V1/DatastoreDbBuilder.cs

```diff
@@ -0,0 +1,247 @@
+        /// ignored.
+        /// </remarks>
+        public EmulatorDetection EmulatorDetection
+        {
+            get => _emulatorDetection;
+            set => _emulatorDetection = GaxPreconditions.CheckEnumValue(value, nameof(value));
+        }
+
+        /// <summary>
+        /// The project ID, or null if this has not yet been configured. Note that a project ID must be configured
+        /// (or detected via an emulator environment variable) before building the <see cref="DatastoreDb"/>.
+        /// </summary>
+        public string ProjectId { get; set; }
+
+        /// <summary>
+        /// Common code for handling the emulator configuration.
+        /// </summary>
+        private ConfiguredBuilder PrepareBuilder()
+        {
+            // TODO: This is slightly awkward, as if a new property is added to ClientBuilderBase, this code will need to be updated.
+            // An alternative would be for Build to end up setting Endpoint and ChannelCredentials on *this* object, which
+            // might also be surprising, and would prevent Build() from being called again. Hmm.
+            // Possible solution: a ClientBuilderBase constructor accepting another ClientBuilderBase, which copies all properties.
+            // We could then have an internal DatastoreClientBuilder constructor accepting a DatastoreDbBuilder.
+            var clientBuilder = new DatastoreClientBuilder
+            {
+                Endpoint = Endpoint,
+                CallInvoker = CallInvoker,
+                ChannelCredentials = ChannelCredentials,
+                CredentialsPath = CredentialsPath,
+                JsonCredentials = JsonCredentials,
+                Scopes = Scopes,
+                Settings = Settings,
+                TokenAccessMethod = TokenAccessMethod
+            };
+            string projectId = ProjectId;
+
+            if (EmulatorDetection != EmulatorDetection.None)
+            {
+                var emulatorConfig = EmulatorConfiguration.FromEnvironment();
+                switch (EmulatorDetection)
+                {
+                    case EmulatorDetection.ProductionOnlyThrowOnEmulatorEnvironment:
+                        GaxPreconditions.CheckState(
+                            emulatorConfig.ProjectId == null && emulatorConfig.Endpoint == null,
+                            "Emulator environment variables are set, contrary to use of {0}.{1}",
+                            nameof(EmulatorDetection), nameof(EmulatorDetection.ProductionOnlyThrowOnEmulatorEnvironment));
+                        break;
+                    case EmulatorDetection.EmulatorOnly:
+                        GaxPreconditions.CheckState(
+                            emulatorConfig.Endpoint != null,
+                            "Expected {0} environment variable to be set", EmulatorConfiguration.EmulatorHostVariable);
+                        ApplyEmulatorConfiguration();
```

Comment:
Does `enumelatorConfig.ProjectId` need checking as not null?

Reply:
Nope - it's okay for just the endpoint to be set, if the project ID has been set in the builder.